File: classification/logit.R
# This script creates and evaluates a multinomial logistic regression model in R for the 'iris' data.
# This iteration is only to get the model up and running, so there is no feature engineering and parameter tuning.

# Load packages.
library(data.table)
library(nnet)

# Import data.
# The data files are located in the 'R' branch file, NOT the 'classification' or 'regression' folders.
train = data.table(read.csv('train.csv',
                            header = T))
test = data.table(read.csv('test.csv', 
                           header = T))
  
#####
# CREATE MODEL AND PREDICTIONS.
  
  # Set seed to ensure reproducibility.
  set.seed(1)
  
  # Define model.
  model = Species ~ Sepal.Length + Sepal.Width + Petal.Length + Petal.Width
  
  # Create model and generate predictions for training set.
  multi_logit = multinom(model, 
              data = train)
  
  # Generate predictions for training and test data.
  train$pred = data.table(predict(multi_logit, 
                                  newdata = train, 
                                  type = 'class'))
  test$pred = data.table(predict(multi_logit, 
                                 newdata = test, 
                                 type = 'class'))

#####
# CHECK TRAINING SET ACCURACY.

  # View training accuracy.
  print('Training Accuracy:')
  table(train$Species == train$pred)
  prop.table(table(train$Species == train$pred))

#####
# CROSS VALIDATION.
# I couldn't find a cross validation function for this technique so I write my own.
  
  # Specify number of folds.
  folds = 2
  
  # Create random row assignment vector.
  train$assignment = sample(1:folds, 
                            replace = T, 
                            nrow(train))
  
  # Create vector to hold cross validation results.
  cv_results = rep(NA, folds)
  
  # Conduct cross validation.
  for (i in 1:folds)
    
    {
    
      # Set training set as all observations with 'assignment == i'.
      train_cv = subset(train, 
                        train$assignment == i)

      # Set test set as all observations with 'assignment != i'.
      test_cv = subset(train, 
                    train$assignment != i)

      # Train model on training data.
      ml = multinom(model, 
                    data = train_cv)
      
      # Generate predictions for test set and append to 'test'.
      test_cv$pred = data.table(predict(ml, 
                                     newdata = test_cv, 
                                     type = 'class'))

      # Add test set error to 'cv_results'.
      cv_results[i] = nrow(subset(test_cv, 
                           test_cv$Species == test_cv$pred)) / nrow(test_cv)
    
    }
  
  # Print cross validation error.
  print('Cross Validation Accuracy:')
  mean(cv_results)
 
#####
# EXPORT TEST SET PREDICTIONS.

  # View test set accuracy.
  print('Test Set Accuracy:')
  table(test$Species == test$pred)
  prop.table(table(test$Species == test$pred))

File: classification/logit.py
# This script creates and evaluates a multinomial logistic regression model in Python for the 'iris' data.
# This iteration is only to get the model up and running, so there is no feature engineering and parameter tuning.

# Import modules.
import pandas as pd
from sklearn import linear_model, cross_validation

# Import data.
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

# CREATE MODEL AND PREDICTIONS.

# Create classifier.
multinom_logit = linear_model.LogisticRegression(random_state = 1)

# Fit classifier on training data.
multinom_logit.fit(train[['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width']],
       train['Species'])

# Generate predictions for training and test data.
train['pred'] = multinom_logit.predict(train[['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width']])
test['pred'] = multinom_logit.predict(test[['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width']])

# CHECK TRAINING SET ACCURACY.

# Compute training set accuracy.
print('Training Set Accuracy :', sum(train['Species'] == train['pred']) / len(train['pred']))

# CROSS VALIDATION.

# Get cross validation scores.
cv_scores = cross_validation.cross_val_score(multinom_logit,
                                             train[['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width']],
                                             train['Species'],
                                             cv = 2)

# Take the mean accuracy across all cross validation segments.
print('Cross Validation Accuracy: ', cv_scores.mean())
                                            
# CHECK TEST SET ACCURACY.

# Compute test set accuracy.
print('Test Set Accuracy :', sum(test['Species'] == test['pred']) / len(test['pred']))
